File: android/core/model/src/main/kotlin/dev/soupy/eclipse/android/core/model/TmdbModels.kt
package dev.soupy.eclipse.android.core.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

private const val TmdbImageBaseUrl = "https://image.tmdb.org/t/p/w780"
private const val TmdbBackdropBaseUrl = "https://image.tmdb.org/t/p/w1280"
private const val TmdbOriginalImageBaseUrl = "https://image.tmdb.org/t/p/original"

@Serializable
data class TMDBSearchResponse(
    val page: Int = 1,
    val results: List<TMDBSearchResult> = emptyList(),
    @SerialName("total_pages") val totalPages: Int = 0,
    @SerialName("total_results") val totalResults: Int = 0,
)

@Serializable
data class TMDBSearchResult(
    val id: Int = 0,
    @SerialName("media_type") val mediaType: String? = null,
    val title: String? = null,
    val name: String? = null,
    val overview: String? = null,
    @SerialName("poster_path") val posterPath: String? = null,
    @SerialName("backdrop_path") val backdropPath: String? = null,
    @SerialName("release_date") val releaseDate: String? = null,
    @SerialName("first_air_date") val firstAirDate: String? = null,
    @SerialName("genre_ids") val genreIds: List<Int> = emptyList(),
    val popularity: Double = 0.0,
)

val TMDBSearchResult.displayTitle: String
    get() = title ?: name ?: "Unknown"

val TMDBSearchResult.displayDate: String?
    get() = releaseDate ?: firstAirDate

val TMDBSearchResult.isMovie: Boolean
    get() = mediaType == "movie" || (mediaType == null && title != null && name == null)

val TMDBSearchResult.isTVShow: Boolean
    get() = mediaType == "tv" || (mediaType == null && name != null && title == null)

val TMDBSearchResult.fullPosterUrl: String?
    get() = posterPath?.let { "$TmdbImageBaseUrl$it" }

val TMDBSearchResult.fullBackdropUrl: String?
    get() = backdropPath?.let { "$TmdbBackdropBaseUrl$it" }

@Serializable
data class TMDBMovie(
    val id: Int = 0,
    val title: String = "",
    val overview: String = "",
    @SerialName("poster_path") val posterPath: String? = null,
    @SerialName("backdrop_path") val backdropPath: String? = null,
    @SerialName("release_date") val releaseDate: String? = null,
    @SerialName("genre_ids") val genreIds: List<Int> = emptyList(),
)

@Serializable
data class TMDBTVShow(
    val id: Int = 0,
    val name: String = "",
    val overview: String = "",
    @SerialName("poster_path") val posterPath: String? = null,
    @SerialName("backdrop_path") val backdropPath: String? = null,
    @SerialName("first_air_date") val firstAirDate: String? = null,
    @SerialName("genre_ids") val genreIds: List<Int> = emptyList(),
)

@Serializable
data class TMDBMovieDetail(
    val id: Int = 0,
    val title: String = "",
    val overview: String = "",
    @SerialName("poster_path") val posterPath: String? = null,
    @SerialName("backdrop_path") val backdropPath: String? = null,
    @SerialName("release_date") val releaseDate: String? = null,
    val runtime: Int? = null,
    val genres: List<TMDBGenre> = emptyList(),
    @SerialName("external_ids") val externalIds: TMDBExternalIds? = null,
)

val TMDBMovieDetail.fullPosterUrl: String?
    get() = posterPath?.let { "$TmdbImageBaseUrl$it" }

val TMDBMovieDetail.fullBackdropUrl: String?
    get() = backdropPath?.let { "$TmdbBackdropBaseUrl$it" }

@Serializable
data class TMDBExternalIds(
    @SerialName("imdb_id") val imdbId: String? = null,
    @SerialName("tvdb_id") val tvdbId: Int? = null,
)

@Serializable
data class TMDBTVShowDetail(
    val id: Int = 0,
    val name: String = "",
    val overview: String = "",
    @SerialName("poster_path") val posterPath: String? = null,
    @SerialName("backdrop_path") val backdropPath: String? = null,
    @SerialName("first_air_date") val firstAirDate: String? = null,
    @SerialName("last_air_date") val lastAirDate: String? = null,
    @SerialName("episode_run_time") val episodeRunTime: List<Int> = emptyList(),
    val genres: List<TMDBGenre> = emptyList(),
    val seasons: List<TMDBSeason> = emptyList(),
    @SerialName("number_of_seasons") val numberOfSeasons: Int? = null,
    @SerialName("number_of_episodes") val numberOfEpisodes: Int? = null,
    val status: String? = null,
    @SerialName("origin_country") val originCountry: List<String> = emptyList(),
    @SerialName("external_ids") val externalIds: TMDBExternalIds? = null,
)

val TMDBTVShowDetail.fullPosterUrl: String?
    get() = posterPath?.let { "$TmdbImageBaseUrl$it" }

val TMDBTVShowDetail.fullBackdropUrl: String?
    get() = backdropPath?.let { "$TmdbBackdropBaseUrl$it" }

@Serializable
data class TMDBGenre(
    val id: Int = 0,
    val name: String = "",
)

@Serializable
data class TMDBSeason(
    val id: Int = 0,
    val name: String = "",
    @SerialName("season_number") val seasonNumber: Int = 0,
    @SerialName("episode_count") val episodeCount: Int = 0,
    @SerialName("poster_path") val posterPath: String? = null,
    @SerialName("air_date") val airDate: String? = null,
)

@Serializable
data class TMDBEpisode(
    val id: Int = 0,
    val name: String = "",
    val overview: String = "",
    @SerialName("episode_number") val episodeNumber: Int = 0,
    @SerialName("season_number") val seasonNumber: Int = 0,
    @SerialName("air_date") val airDate: String? = null,
    @SerialName("runtime") val runtime: Int? = null,
    @SerialName("still_path") val stillPath: String? = null,
)

val TMDBEpisode.fullStillUrl: String?
    get() = stillPath?.let { "$TmdbImageBaseUrl$it" }

@Serializable
data class TMDBSeasonDetail(
    val id: Int = 0,
    val name: String = "",
    val overview: String = "",
    @SerialName("season_number") val seasonNumber: Int = 0,
    @SerialName("air_date") val airDate: String? = null,
    @SerialName("poster_path") val posterPath: String? = null,
    val episodes: List<TMDBEpisode> = emptyList(),
)

@Serializable
data class TMDBCreditsResponse(
    val id: Int = 0,
    val cast: List<TMDBCastMember> = emptyList(),
)

@Serializable
data class TMDBCastMember(
    val id: Int = 0,
    val name: String = "",
    val character: String = "",
    @SerialName("profile_path") val profilePath: String? = null,
    val order: Int = 0,
)

val TMDBCastMember.fullProfileUrl: String?
    get() = profilePath?.let { "$TmdbImageBaseUrl$it" }

@Serializable
data class TMDBContentRatingsResponse(
    val results: List<TMDBContentRating> = emptyList(),
)

@Serializable
data class TMDBContentRating(
    @SerialName("iso_3166_1") val countryCode: String = "",
    val rating: String = "",
)

val TMDBContentRatingsResponse.usRating: String?
    get() = results.firstOrNull { it.countryCode.equals("US", ignoreCase = true) && it.rating.isNotBlank() }
        ?.rating
        ?: results.firstOrNull { it.rating.isNotBlank() }?.rating

@Serializable
data class TMDBReleaseDatesResponse(
    val results: List<TMDBReleaseDateCountry> = emptyList(),
)

@Serializable
data class TMDBReleaseDateCountry(
    @SerialName("iso_3166_1") val countryCode: String = "",
    @SerialName("release_dates") val releaseDates: List<TMDBReleaseDateEntry> = emptyList(),
)

@Serializable
data class TMDBReleaseDateEntry(
    val certification: String = "",
    val type: Int = 0,
)

val TMDBReleaseDatesResponse.usCertification: String?
    get() {
        val usCertification = results
            .firstOrNull { it.countryCode.equals("US", ignoreCase = true) }
            ?.releaseDates
            ?.firstOrNull { it.certification.isNotBlank() }
            ?.certification
        return usCertification
            ?: results
                .flatMap { it.releaseDates }
                .firstOrNull { it.certification.isNotBlank() }
                ?.certification
    }

@Serializable
data class TMDBImagesResponse(
    val id: Int = 0,
    val backdrops: List<TMDBImage>? = null,
    val logos: List<TMDBImage>? = null,
    val posters: List<TMDBImage>? = null,
)

@Serializable
data class TMDBImage(
    @SerialName("aspect_ratio") val aspectRatio: Double = 0.0,
    val height: Int = 0,
    val width: Int = 0,
    @SerialName("file_path") val filePath: String = "",
    @SerialName("iso_639_1") val languageCode: String? = null,
    @SerialName("vote_average") val voteAverage: Double? = null,
    @SerialName("vote_count") val voteCount: Int? = null,
)

val TMDBImage.fullOriginalUrl: String?
    get() = filePath.takeIf { it.isNotBlank() }?.let { "$TmdbOriginalImageBaseUrl$it" }

fun TMDBImagesResponse.bestLogoUrl(preferredLanguage: String?): String? {
    val availableLogos = logos.orEmpty()
    if (availableLogos.isEmpty()) return null
    val languagePrefix = preferredLanguage
        ?.substringBefore('-')
        ?.takeIf { it.isNotBlank() }
        ?: "en"
    return availableLogos.firstOrNull { it.languageCode == languagePrefix }?.fullOriginalUrl
        ?: availableLogos.firstOrNull { it.languageCode == "en" }?.fullOriginalUrl
        ?: availableLogos.firstOrNull { it.languageCode == null }?.fullOriginalUrl
        ?: availableLogos.firstOrNull()?.fullOriginalUrl
}

@Serializable
data class TMDBTVShowWithSeasons(
    val show: TMDBTVShowDetail,
    val seasonDetails: List<TMDBSeasonDetail> = emptyList(),
)

File: Luna/TMDB/Models/TMDBModels.swift
//
//  TMDBModels.swift
//  Sora
//
//  Created by Francesco on 07/08/25.
//

import Foundation

private struct LossyDecodableArray<Element: Decodable>: Decodable {
    let elements: [Element]
    let skippedCount: Int

    init(from decoder: Decoder) throws {
        var container = try decoder.unkeyedContainer()
        var elements: [Element] = []
        var skippedCount = 0

        while !container.isAtEnd {
            let value = try container.decode(LossyDecodableValue<Element>.self)
            if let element = value.element {
                elements.append(element)
            } else {
                skippedCount += 1
            }
        }

        self.elements = elements
        self.skippedCount = skippedCount
    }
}

private struct LossyDecodableValue<Element: Decodable>: Decodable {
    let element: Element?

    init(from decoder: Decoder) throws {
        element = try? Element(from: decoder)
    }
}

// MARK: - Search Response
struct TMDBSearchResponse: Decodable {
    let page: Int
    let results: [TMDBSearchResult]
    let totalPages: Int
    let totalResults: Int
    let skippedResultCount: Int
    
    enum CodingKeys: String, CodingKey {
        case page, results
        case totalPages = "total_pages"
        case totalResults = "total_results"
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let lossyResults = try container.decodeIfPresent(LossyDecodableArray<TMDBSearchResult>.self, forKey: .results)

        page = try container.decodeIfPresent(Int.self, forKey: .page) ?? 1
        results = lossyResults?.elements ?? []
        totalPages = try container.decodeIfPresent(Int.self, forKey: .totalPages) ?? 1
        totalResults = try container.decodeIfPresent(Int.self, forKey: .totalResults) ?? results.count
        skippedResultCount = lossyResults?.skippedCount ?? 0
    }

    init(page: Int, results: [TMDBSearchResult], totalPages: Int, totalResults: Int, skippedResultCount: Int = 0) {
        self.page = page
        self.results = results
        self.totalPages = totalPages
        self.totalResults = totalResults
        self.skippedResultCount = skippedResultCount
    }
}

struct TMDBFindResponse: Decodable {
    let movieResults: [TMDBMovie]
    let tvResults: [TMDBTVShow]

    enum CodingKeys: String, CodingKey {
        case movieResults = "movie_results"
        case tvResults = "tv_results"
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        movieResults = (try container.decodeIfPresent(LossyDecodableArray<TMDBMovie>.self, forKey: .movieResults))?.elements ?? []
        tvResults = (try container.decodeIfPresent(LossyDecodableArray<TMDBTVShow>.self, forKey: .tvResults))?.elements ?? []
    }
}

// MARK: - Search Result
struct TMDBSearchResult: Codable, Identifiable {
    let id: Int
    let mediaType: String
    let title: String?
    let name: String?
    let overview: String?
    let posterPath: String?
    let backdropPath: String?
    let releaseDate: String?
    let firstAirDate: String?
    let voteAverage: Double?
    let popularity: Double
    let adult: Bool?
    let genreIds: [Int]?
    
    enum CodingKeys: String, CodingKey {
        case id, overview, popularity, adult
        case mediaType = "media_type"
        case title, name
        case posterPath = "poster_path"
        case backdropPath = "backdrop_path"
        case releaseDate = "release_date"
        case firstAirDate = "first_air_date"
        case voteAverage = "vote_average"
        case genreIds = "genre_ids"
    }
    
    var displayTitle: String {
        return title ?? name ?? "Unknown Title"
    }
    
    var displayDate: String {
        return releaseDate ?? firstAirDate ?? ""
    }
    
    var isMovie: Bool {
        return mediaType == "movie"
    }
    
    var isTVShow: Bool {
        return mediaType == "tv"
    }
    
    var fullPosterURL: String? {
        guard let posterPath = posterPath else { return nil }
        if posterPath.lowercased().hasPrefix("http://") || posterPath.lowercased().hasPrefix("https://") {
            return posterPath
        }
        return "\(TMDBService.tmdbImageBaseURL)\(posterPath)"
    }
    
    var fullBackdropURL: String? {
        guard let backdropPath = backdropPath else { return nil }
        if backdropPath.lowercased().hasPrefix("http://") || backdropPath.lowercased().hasPrefix("https://") {
            return backdropPath
        }
        return "\(TMDBService.tmdbImageBaseURL)\(backdropPath)"
    }

    var stableIdentity: String {
        "\(mediaType)-\(id)"
    }
}

// MARK: - Movie Search Response
struct TMDBMovieSearchResponse: Decodable {
    let page: Int
    let results: [TMDBMovie]
    let totalPages: Int
    let totalResults: Int
    let skippedResultCount: Int
    
    enum CodingKeys: String, CodingKey {
        case page, results
        case totalPages = "total_pages"
        case totalResults = "total_results"
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let lossyResults = try container.decodeIfPresent(LossyDecodableArray<TMDBMovie>.self, forKey: .results)

        page = try container.decodeIfPresent(Int.self, forKey: .page) ?? 1
        results = lossyResults?.elements ?? []
        totalPages = try container.decodeIfPresent(Int.self, forKey: .totalPages) ?? 1
        totalResults = try container.decodeIfPresent(Int.self, forKey: .totalResults) ?? results.count
        skippedResultCount = lossyResults?.skippedCount ?? 0
    }

    init(page: Int, results: [TMDBMovie], totalPages: Int, totalResults: Int, skippedResultCount: Int = 0) {
        self.page = page
        self.results = results
        self.totalPages = totalPages
        self.totalResults = totalResults
        self.skippedResultCount = skippedResultCount
    }
}

// MARK: - TV Show Search Response
struct TMDBTVSearchResponse: Decodable {
    let page: Int
    let results: [TMDBTVShow]
    let totalPages: Int
    let totalResults: Int
    let skippedResultCount: Int
    
    enum CodingKeys: String, CodingKey {
        case page, results
        case totalPages = "total_pages"
        case totalResults = "total_results"
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let lossyResults = try container.decodeIfPresent(LossyDecodableArray<TMDBTVShow>.self, forKey: .results)

        page = try container.decodeIfPresent(Int.self, forKey: .page) ?? 1
        results = lossyResults?.elements ?? []
        totalPages = try container.decodeIfPresent(Int.self, forKey: .totalPages) ?? 1
        totalResults = try container.decodeIfPresent(Int.self, forKey: .totalResults) ?? results.count
        skippedResultCount = lossyResults?.skippedCount ?? 0
    }

    init(page: Int, results: [TMDBTVShow], totalPages: Int, totalResults: Int, skippedResultCount: Int = 0) {
        self.page = page
        self.results = results
        self.totalPages = totalPages
        self.totalResults = totalResults
        self.skippedResultCount = skippedResultCount
    }
}

// MARK: - Movie Model
struct TMDBMovie: Codable, Identifiable {
    let id: Int
    let title: String
    let overview: String?
    let posterPath: String?
    let backdropPath: String?
    let releaseDate: String?
    let voteAverage: Double
    let popularity: Double
    let adult: Bool?
    let genreIds: [Int]?
    
    enum CodingKeys: String, CodingKey {
        case id, title, overview, popularity, adult
        case posterPath = "poster_path"
        case backdropPath = "backdrop_path"
        case releaseDate = "release_date"
        case voteAverage = "vote_average"
        case genreIds = "genre_ids"
    }
    
    var fullPosterURL: String? {
        guard let posterPath = posterPath else { return nil }
        return "\(TMDBService.tmdbImageBaseURL)\(posterPath)"
    }
    
    var fullBackdropURL: String? {
        guard let backdropPath = backdropPath else { return nil }
        return "\(TMDBService.tmdbImageBaseURL)\(backdropPath)"
    }
    
    var asSearchResult: TMDBSearchResult {
        return TMDBSearchResult(
            id: id,
            mediaType: "movie",
            title: title,
            name: nil,
            overview: overview,
            posterPath: posterPath,
            backdropPath: backdropPath,
            releaseDate: releaseDate,
            firstAirDate: nil,
            voteAverage: voteAverage,
            popularity: popularity,
            adult: adult,
            genreIds: genreIds
        )
    }
}

// MARK: - TV Show Model
struct TMDBTVShow: Codable, Identifiable {
    let id: Int
    let name: String
    let overview: String?
    let posterPath: String?
    let backdropPath: String?
    let firstAirDate: String?
    let voteAverage: Double
    let popularity: Double
    let genreIds: [Int]?
    
    enum CodingKeys: String, CodingKey {
        case id, name, overview, popularity
        case posterPath = "poster_path"
        case backdropPath = "backdrop_path"
        case firstAirDate = "first_air_date"
        case voteAverage = "vote_average"
        case genreIds = "genre_ids"
    }
    
    var fullPosterURL: String? {
        guard let posterPath = posterPath else { return nil }
        return "\(TMDBService.tmdbImageBaseURL)\(posterPath)"
    }
    
    var fullBackdropURL: String? {
        guard let backdropPath = backdropPath else { return nil }
        return "\(TMDBService.tmdbImageBaseURL)\(backdropPath)"
    }
    
    var asSearchResult: TMDBSearchResult {
        return TMDBSearchResult(
            id: id,
            mediaType: "tv",
            title: nil,
            name: name,
            overview: overview,
            posterPath: posterPath,
            backdropPath: backdropPath,
            releaseDate: nil,
            firstAirDate: firstAirDate,
            voteAverage: voteAverage,
            popularity: popularity,
            adult: nil,
            genreIds: genreIds
        )
    }
}

// MARK: - Movie Detail Model
struct TMDBMovieDetail: Codable, Identifiable {
    let id: Int
    let title: String
    let overview: String?
    let posterPath: String?
    let backdropPath: String?
    let releaseDate: String?
    let voteAverage: Double
    let popularity: Double
    let runtime: Int?
    let genres: [TMDBGenre]
    let tagline: String?
    let status: String?
    let budget: Int?
    let revenue: Int?
    let imdbId: String?
    let originalLanguage: String?
    let originalTitle: String?
    let adult: Bool
    let voteCount: Int
    let releaseDates: TMDBReleaseDates?
    
    enum CodingKeys: String, CodingKey {
        case id, title, overview, popularity, runtime, genres, tagline, status, budget, revenue, adult
        case posterPath = "poster_path"
        case backdropPath = "backdrop_path"
        case releaseDate = "release_date"
        case voteAverage = "vote_average"
        case imdbId = "imdb_id"
        case originalLanguage = "original_language"
        case originalTitle = "original_title"
        case voteCount = "vote_count"
        case releaseDates = "release_dates"
    }
    
    var fullPosterURL: String? {
        guard let posterPath = posterPath else { return nil }
        return "\(TMDBService.tmdbImageBaseURL)\(posterPath)"
    }
    
    var fullBackdropURL: String? {
        guard let backdropPath = backdropPath else { return nil }
        return "\(TMDBService.tmdbImageBaseURL)\(backdropPath)"
    }
    
    var runtimeFormatted: String {
        guard let runtime = runtime, runtime > 0 else { return "Unknown" }
        let hours = runtime / 60
        let minutes = runtime % 60
        if hours > 0 {
            return "\(hours)h \(minutes)m"
        } else {
            return "\(minutes)m"
        }
    }
    
    var yearFromReleaseDate: String {
        guard let releaseDate = releaseDate, !releaseDate.isEmpty else { return "Unknown" }
        return String(releaseDate.prefix(4))
    }
}

// MARK: - External IDs (for IMDB lookup)
struct TMDBExternalIds: Codable {
    let imdbId: String?
    let freebaseMid: String?
    let freebaseId: String?
    let tvdbId: Int?
    let tvrageId: Int?
    let facebookId: String?
    let instagramId: String?
    let twitterId: String?

    enum CodingKeys: String, CodingKey {
        case imdbId = "imdb_id"
        case freebaseMid = "freebase_mid"
        case freebaseId = "freebase_id"
        case tvdbId = "tvdb_id"
        case tvrageId = "tvrage_id"
        case facebookId = "facebook_id"
        case instagramId = "instagram_id"
        case twitterId = "twitter_id"
    }
}

// MARK: - TV Show Detail Model
struct TMDBTVShowDetail: Codable, Identifiable {
    let id: Int
    let name: String
    let overview: String?
    let posterPath: String?
    let backdropPath: String?
    let firstAirDate: String?
    let lastAirDate: String?
    let voteAverage: Double
    let popularity: Double
    let genres: [TMDBGenre]
    let tagline: String?
    let status: String?
    let originalLanguage: String?
    let originalName: String?
    let adult: Bool
    let voteCount: Int
    let numberOfSeasons: Int?
    let numberOfEpisodes: Int?
    let episodeRunTime: [Int]?
    let inProduction: Bool?
    let languages: [String]?
    let originCountry: [String]?
    let type: String?
    let contentRatings: TMDBContentRatings?
    let externalIds: TMDBExternalIds?
    let nextEpisodeToAir: TMDBEpisode?
    
    enum CodingKeys: String, CodingKey {
        case id, name, overview, popularity, genres, tagline, status, adult, languages, type
        case posterPath = "poster_path"
        case backdropPath = "backdrop_path"
        case firstAirDate = "first_air_date"
        case lastAirDate = "last_air_date"
        case voteAverage = "vote_average"
        case originalLanguage = "original_language"
        case originalName = "original_name"
        case voteCount = "vote_count"
        case numberOfSeasons = "number_of_seasons"
        case numberOfEpisodes = "number_of_episodes"
        case episodeRunTime = "episode_run_time"
        case inProduction = "in_production"
        case originCountry = "origin_country"
        case contentRatings = "content_ratings"
        case externalIds = "external_ids"
        case nextEpisodeToAir = "next_episode_to_air"
    }
    
    var fullPosterURL: String? {
        guard let posterPath = posterPath else { return nil }
        return "\(TMDBService.tmdbImageBaseURL)\(posterPath)"
    }
    
    var fullBackdropURL: String? {
        guard let backdropPath = backdropPath else { return nil }
        return "\(TMDBService.tmdbImageBaseURL)\(backdropPath)"
    }
    
    var yearFromFirstAirDate: String {
        guard let firstAirDate = firstAirDate, !firstAirDate.isEmpty else { return "Unknown" }
        return String(firstAirDate.prefix(4))
    }
    
    var episodeRuntimeFormatted: String {
        guard let runtime = episodeRunTime?.first, runtime > 0 else { return "Unknown" }
        return "\(runtime)m"
    }
}

// MARK: - Genre Model
struct TMDBGenre: Codable, Identifiable {
    let id: Int
    let name: String
}

// MARK: - Season Model
struct TMDBSeason: Codable, Identifiable {
    let id: Int
    let name: String
    let overview: String?
    let posterPath: String?
    let seasonNumber: Int
    let episodeCount: Int
    let airDate: String?
    
    enum CodingKeys: String, CodingKey {
        case id, name, overview
        case posterPath = "poster_path"
        case seasonNumber = "season_number"
        case episodeCount = "episode_count"
        case airDate = "air_date"
    }
    
    var fullPosterURL: String? {
        guard let posterPath = posterPath else { return nil }
        // If we already have a full URL (e.g., AniList CDN), return it directly
        if posterPath.hasPrefix("http") { return posterPath }
        return "\(TMDBService.tmdbImageBaseURL)\(posterPath)"
    }
}

// MARK: - Episode Model
struct TMDBEpisode: Codable, Identifiable {
    let id: Int
    let name: String
    let overview: String?
    let stillPath: String?
    let episodeNumber: Int
    let seasonNumber: Int
    let airDate: String?
    let runtime: Int?
    let voteAverage: Double
    let voteCount: Int
    
    enum CodingKeys: String, CodingKey {
        case id, name, overview, runtime
        case stillPath = "still_path"
        case episodeNumber = "episode_number"
        case seasonNumber = "season_number"
        case airDate = "air_date"
        case voteAverage = "vote_average"
        case voteCount = "vote_count"
    }
    
    var fullStillURL: String? {
        guard let stillPath = stillPath else { return nil }
        return "\(TMDBService.tmdbImageBaseURL)\(stillPath)"
    }
    
    var runtimeFormatted: String {
        guard let runtime = runtime, runtime > 0 else { return "Unknown" }
        return "\(runtime)m"
    }
}

// MARK: - Season Detail Model
struct TMDBSeasonDetail: Codable, Identifiable {
    let id: Int
    let name: String
    let overview: String?
    let posterPath: String?
    let seasonNumber: Int
    let airDate: String?
    let episodes: [TMDBEpisode]
    
    enum CodingKeys: String, CodingKey {
        case id, name, overview, episodes
        case posterPath = "poster_path"
        case seasonNumber = "season_number"
        case airDate = "air_date"
    }
    
    var fullPosterURL: String? {
        guard let posterPath = posterPath else { return nil }
        if posterPath.hasPrefix("http") { return posterPath }
        return "\(TMDBService.tmdbImageBaseURL)\(posterPath)"
    }
}

// MARK: - TV Show with Seasons
struct TMDBTVShowWithSeasons: Codable, Identifiable {
    let id: Int
    let name: String
    let overview: String?
    let posterPath: String?
    let backdropPath: String?
    let firstAirDate: String?
    let lastAirDate: String?
    let voteAverage: Double
    let popularity: Double
    let genres: [TMDBGenre]
    let tagline: String?
    let status: String?
    let originalLanguage: String?
    let originalName: String?
    let adult: Bool
    let voteCount: Int
    let numberOfSeasons: Int?
    let numberOfEpisodes: Int?
    let episodeRunTime: [Int]?
    let inProduction: Bool?
    let languages: [String]?
    let originCountry: [String]?
    let type: String?
    let seasons: [TMDBSeason]
    let contentRatings: TMDBContentRatings?
    let externalIds: TMDBExternalIds?
    
    enum CodingKeys: String, CodingKey {
        case id, name, overview, popularity, genres, tagline, status, adult, languages, type, seasons
        case posterPath = "poster_path"
        case backdropPath = "backdrop_path"
        case firstAirDate = "first_air_date"
        case lastAirDate = "last_air_date"
        case voteAverage = "vote_average"
        case originalLanguage = "original_language"
        case originalName = "original_name"
        case voteCount = "vote_count"
        case numberOfSeasons = "number_of_seasons"
        case numberOfEpisodes = "number_of_episodes"
        case episodeRunTime = "episode_run_time"
        case inProduction = "in_production"
        case originCountry = "origin_country"
        case contentRatings = "content_ratings"
        case externalIds = "external_ids"
    }
    
    var fullPosterURL: String? {
        guard let posterPath = posterPath else { return nil }
        return "\(TMDBService.tmdbImageBaseURL)\(posterPath)"
    }
    
    var fullBackdropURL: String? {
        guard let backdropPath = backdropPath else { return nil }
        return "\(TMDBService.tmdbImageBaseURL)\(backdropPath)"
    }
    
    var yearFromFirstAirDate: String {
        guard let firstAirDate = firstAirDate, !firstAirDate.isEmpty else { return "Unknown" }
        return String(firstAirDate.prefix(4))
    }
    
    var episodeRuntimeFormatted: String {
        guard let runtime = episodeRunTime?.first, runtime > 0 else { return "Unknown" }
        return "\(runtime)m"
    }
}

// MARK: - Alternative Titles
struct TMDBAlternativeTitles: Codable {
    let id: Int
    let titles: [TMDBAlternativeTitle]
}

struct TMDBAlternativeTitle: Codable {
    let iso31661: String
    let title: String
    let type: String?
    
    enum CodingKeys: String, CodingKey {
        case title, type
        case iso31661 = "iso_3166_1"
    }
}

// MARK: - TV Alternative Titles
struct TMDBTVAlternativeTitles: Codable {
    let id: Int
    let results: [TMDBTVAlternativeTitle]
}

struct TMDBTVAlternativeTitle: Codable {
    let iso31661: String
    let title: String
    let type: String?
    
    enum CodingKeys: String, CodingKey {
        case title, type
        case iso31661 = "iso_3166_1"
    }
}

// MARK: - Content Ratings Models
struct TMDBReleaseDates: Codable {
    let results: [TMDBReleaseDateResult]
}

struct TMDBReleaseDateResult: Codable {
    let iso31661: String
    let releaseDates: [TMDBReleaseDate]
    
    enum CodingKeys: String, CodingKey {
        case iso31661 = "iso_3166_1"
        case releaseDates = "release_dates"
    }
}

struct TMDBReleaseDate: Codable {
    let certification: String
    let iso6391: String?
    let note: String?
    let releaseDate: String
    let type: Int
    
    enum CodingKeys: String, CodingKey {
        case certification, note, type
        case iso6391 = "iso_639_1"
        case releaseDate = "release_date"
    }
}

struct TMDBContentRatings: Codable {
    let results: [TMDBContentRating]
}

struct TMDBContentRating: Codable {
    let descriptors: [String]?
    let iso31661: String
    let rating: String
    
    enum CodingKeys: String, CodingKey {
        case descriptors, rating
        case iso31661 = "iso_3166_1"
    }
}

// MARK: - Images Response
struct TMDBImagesResponse: Codable {
    let id: Int
    let backdrops: [TMDBImage]?
    let logos: [TMDBImage]?
    let posters: [TMDBImage]?
}

struct TMDBImage: Codable {
    let aspectRatio: Double
    let height: Int
    let width: Int
    let filePath: String
    let iso6391: String?
    let voteAverage: Double?
    let voteCount: Int?
    
    enum CodingKeys: String, CodingKey {
        case height, width
        case aspectRatio = "aspect_ratio"
        case filePath = "file_path"
        case iso6391 = "iso_639_1"
        case voteAverage = "vote_average"
        case voteCount = "vote_count"
    }
    
    var fullURL: String {
        return "\(TMDBService.tmdbImageBaseURL)\(filePath)"
    }
}

// MARK: - Credits / Cast
struct TMDBCreditsResponse: Codable {
    let id: Int
    let cast: [TMDBCastMember]
}

struct TMDBCastMember: Codable, Identifiable {
    let id: Int
    let name: String
    let character: String?
    let profilePath: String?
    let order: Int?
    let knownForDepartment: String?
    let popularity: Double?
    
    enum CodingKeys: String, CodingKey {
        case id, name, character, order, popularity
        case profilePath = "profile_path"
        case knownForDepartment = "known_for_department"
    }
    
    var fullProfileURL: String? {
        guard let profilePath = profilePath else { return nil }
        return "\(TMDBService.tmdbImageBaseURL)\(profilePath)"
    }
}
